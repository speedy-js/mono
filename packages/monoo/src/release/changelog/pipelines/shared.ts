export type TChangelogProcessorOptions = object;

export type TChangelogProcessor<
  T extends TChangelogProcessorOptions = TChangelogProcessorOptions,
> = (cwd: string, input: string, options: T) => string;

export type TChangelogProcessorTuple = [
  TChangelogProcessor,
  TChangelogProcessorOptions,
];

export class ChangelogProcessor {
  constructor(private cwd: string, private input: string) {}

  private processorsTuple: TChangelogProcessorTuple[] = [];

  /**
   * Add a processor
   */
  addProcessor<T extends TChangelogProcessor>(
    processor: T,
    options: Parameters<T>['2'],
  ) {
    this.processorsTuple.push([processor, options]);
  }

  /**
   * Process and get final result
   */
  process() {
    const { cwd, input } = this;
    return this.processorsTuple.reduce<string>((memo, current) => {
      memo = current[0](cwd, memo, current[1]);
      return memo;
    }, input);
  }
}
