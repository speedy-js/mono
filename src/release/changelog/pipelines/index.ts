export * from './beautify';
export * from './attach-author';
export * from './shared';
export * from './normalize-commit-url';
