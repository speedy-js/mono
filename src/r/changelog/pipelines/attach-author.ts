/**
 * Module dependencies
 */
import { ChangelogNS } from '../../types';
import { TChangelogProcessor } from './shared';
import { getCommitHashAuthorMap } from '../helpers';
import { COMMIT_RE } from '../regexp';

export interface IDisplayAuthorOptions {
  getAuthorPage?(author: ChangelogNS.ICommitAuthor): string;
  displayType?: 'name' | 'email';
}

export const ChangelogProcessor4AttachAuthor: TChangelogProcessor<IDisplayAuthorOptions> =
  (cwd, input, options = {}) => {
    const displayType = options.displayType ?? 'name';
    const getAuthorPage =
      options.getAuthorPage ??
      ((author: ChangelogNS.ICommitAuthor) => {
        return `https://github.com/${author.emailName}`;
      });

    const hashAuthorMap = getCommitHashAuthorMap(cwd);
    const ret = input.replace(
      COMMIT_RE,
      (
        m,
        s1 /* hash */,
        s2 /* url */,
        s3 /* white space */,
        s4 /* author part */,
        s5 /* author text */,
      ): string => {
        const author = hashAuthorMap[s1];
        /**
         * s4 exists, do not need generate author
         */
        if (!author || s4) {
          return m;
        }
        const dispalyName =
          displayType === 'name' ? author.name : author.emailName;
        const authorUrl = getAuthorPage(author);
        return `([${s1}](${s2})) [@${dispalyName}](${authorUrl})`;
      },
    );

    return ret;
  };
