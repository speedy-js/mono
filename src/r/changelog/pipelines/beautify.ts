/**
 * Module dependencies
 */
import { TChangelogProcessor } from './shared';

/**
 * Remove top-level useless header
 *
 * @see https://github.com/conventional-changelog/conventional-changelog/issues/376
 * @param cwd
 * @param input
 * @returns
 */
export const ChangelogProcessor4Beautify: TChangelogProcessor = (
  cwd,
  input,
) => {
  input = input.trim();
  const changelogLines = input.split('\n');
  if (/# \[((0|1)\.0\.0)?\]/.test(changelogLines[0])) {
    input = changelogLines.slice(1).join('\n').trim();
  }
  return input;
};
