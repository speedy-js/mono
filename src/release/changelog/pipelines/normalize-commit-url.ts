/**
 * Module dependencies
 */
import { TChangelogProcessor } from './shared';
import { COMMIT_RE } from '../regexp';

/**
 * Normalize commit url
 *
 * @param cwd
 * @param input
 * @returns
 */
export const ChangelogProcessor4NormalizeCommitUrl: TChangelogProcessor = (
  cwd,
  input,
) => {
  return input.replace(
    COMMIT_RE,
    (
      m,
      s1 /* hash */,
      s2 /* url */,
      s3 /* white space */,
      s4 /* author part */,
      s5 /* author text */,
    ): string => {
      return m.replace('commits', 'commit');
    },
  );
};
