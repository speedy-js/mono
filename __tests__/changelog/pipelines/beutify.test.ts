/**
 * Module dependencies
 */
import { ChangelogProcessor4Beautify } from '../../../src/r/changelog/pipelines/beautify';

describe('ChangelogProcessor4Beautify', () => {

  it('remove 0.0.0', () => {
    const input = `# [0.0.0](https://github.com/speedy/mono/compare/v0.2.3...v0.2.4) (2021-10-23)

    ## [0.2.4](https://github.com/speedy/mono/compare/v0.2.3...v0.2.4) (2021-10-23)
    
    
    ### Bug Fixes
    
    * \`--ignore-scripts\` description ([663a396](https://github.com/speedy/mono/commits/663a3963b63706d273068895f4dd79d6c70472c2))
      
      `;

    expect(ChangelogProcessor4Beautify('<cwd>', input, {})).toMatchSnapshot();
  });
  
  it('remove 1.0.0', () => {
    const input = `# [1.0.0](https://github.com/foo/bar/compare/v0.5.22...v1.0.0) (2021-11-24)



## [0.5.22](https://github.com/foo/bar/compare/v0.5.21...v0.5.22) (2021-11-24)
    
    `
    expect(ChangelogProcessor4Beautify('<cwd>', input, {})).toMatchSnapshot();
  }); 
  
  it('remove empty header', () => {
    const input = `# [](https://github.com/foo/bar/compare/v0.1.1...v) (2021-12-04)



## [0.1.1](https://github.com/foo/bar/compare/v0.1.0...v0.1.1) (2021-12-04)
    `
    expect(ChangelogProcessor4Beautify('<cwd>', input, {})).toMatchSnapshot();
  });
});
