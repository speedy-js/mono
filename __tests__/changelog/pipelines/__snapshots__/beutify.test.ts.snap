// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChangelogProcessor4Beautify remove 0.0.0 1`] = `
"## [0.2.4](https://github.com/speedy/mono/compare/v0.2.3...v0.2.4) (2021-10-23)
    
    
    ### Bug Fixes
    
    * \`--ignore-scripts\` description ([663a396](https://github.com/speedy/mono/commits/663a3963b63706d273068895f4dd79d6c70472c2))"
`;

exports[`ChangelogProcessor4Beautify remove 1.0.0 1`] = `"## [0.5.22](https://github.com/foo/bar/compare/v0.5.21...v0.5.22) (2021-11-24)"`;

exports[`ChangelogProcessor4Beautify remove empty header 1`] = `"## [0.1.1](https://github.com/foo/bar/compare/v0.1.0...v0.1.1) (2021-12-04)"`;
